### ISSUE TITLE
bug: require-error checker can break code while it shouldn't

### ISSUE THREAD

**Issue by ccoVeille** [Antonboom/testifylint#79]
Here is the checker

https://github.com/Antonboom/testifylint#require-error
```
❌   assert.Error(t, err) // s.Error(err), s.Assert().Error(err)

✅   require.Error(t, err) // s.Require().Error(err), s.Require().Error(err)
     // And so on...
```

There is one small issue with this linter based on an assumption: t *testing.T implements everything.

But here are the interfaces testify accept:
https://pkg.go.dev/github.com/stretchr/testify/assert#TestingT

```go
type TestingT interface {
	Errorf(format [string](https://pkg.go.dev/builtin#string), args ...interface{})
}
```

https://pkg.go.dev/github.com/stretchr/testify/require#TestingT

```go
type TestingT interface {
	Errorf(format [string](https://pkg.go.dev/builtin#string), args ...interface{})
	FailNow()
}
```

So when `require-error` does it check it MUST check that the `t` implements `testify/require/TestingT`

Right now, it doesn't so it break code.

https://github.com/Boeing/config-file-validator/blob/9006fcef0d9ac5587f516d7181454b699758509e/pkg/reporter/reporter_test.go#L375

```go
func assertRegexpError(regexp interface{}) assert.ErrorAssertionFunc {
	return func(t assert.TestingT, got error, msg ...interface{}) bool {
		if h, ok := t.(interface{ Helper() }); ok {
			h.Helper()
		}
		return assert.Error(t, got, msg...) && assert.Regexp(t, regexp, got.Error(), msg...)
	}
}
```

Here the problem is with 

https://pkg.go.dev/github.com/stretchr/testify/assert#ErrorAssertionFunc
https://pkg.go.dev/github.com/stretchr/testify/require#ErrorAssertionFunc

that relies on `assert.TestingT` vs `require.TestingT`

**Comment by ccoVeille** [comment]
It's definitely an edge case, but I think the priority is to fix the wrong suggestion.
There is no autofix for this checker, so it's a good thing.

We could have a discussion about how to report such issue.

**Comment by ccoVeille** [comment]
For records, I'm close to a "fix"

**Comment by Antonboom** [comment]
Oh, `require-error` is painful checker with very complex implementation :) 

>  So when require-error does it check it MUST check that the t implements testify/require/TestingT

Fix looks pretty simple: 

```
133 testifyCall := NewCallMeta(pass, ce)
134 if isRequireTestingTPtr(testifyCall) { ... }
```

where `isRequireTestingTPtr` is part of `isTestingTPtr`

> For records, I'm close to a "fix"

Do not forget to add tests before fix (TDD) 👌

**Comment by ccoVeille** [comment]
Yes, that's exactly what I tried, hand yes, I'm facing the suite detection issue.

IsPkG helped me but it didn't fix everything.

At some point, I will post my WIP code, and ask for help, I think

**Comment by Antonboom** [comment]
> I will post my WIP code, and ask for help, I think

welcome

**Comment by ccoVeille** [comment]
I just became dad again. My priorities changed. I'll be busy before coming back to dev.

**Comment by Antonboom** [comment]
@ccoVeille 

Having studied your case in more detail, I decided that it would be wrong to move the checker only to `require.TestingT` interface.

In such case, code like this can be ignored:

```go
func assertRegexpError2(regexp any) assert.ErrorAssertionFunc {
	return func(t assert.TestingT, got error, msg ...any) bool {
		if h, ok := t.(interface{ Helper() }); ok {
			h.Helper()
		}
		assert.Error(t, got, msg...) // want "require-error: for error assertions use require"
		return assert.Regexp(t, regexp, got.Error(), msg...)
	}
}
```

But! I think the problem here is different – just assertions in boolean expressions should be ignored, which is what I did.

See related pr, thanks!

**Comment by ccoVeille** [comment]
I'm fine with your change.

Anyway, the checker is not providing an autofix, so it won't break code.

And if the asserter available is not providing the FailNow, so what is needed to implement require.T, then suggesting to use a require is OK. Either the asserter should implement the FailNow, or pass another asserter
